File: benchmarks/pandas/bench_dataframe_rolling_var_std_sum_count.py
"""
Benchmark: pandas DataFrame.rolling().var() / std() / sum() / count() — rolling aggregations.
Outputs JSON: {"function": "dataframe_rolling_var_std_sum_count", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import time
import pandas as pd
import numpy as np

SIZE = 50_000
WINDOW = 20
WARMUP = 5
ITERATIONS = 30

df = pd.DataFrame({
    "a": np.sin(np.arange(SIZE) * 0.01) * 100,
    "b": np.cos(np.arange(SIZE) * 0.01) * 50,
    "c": (np.arange(SIZE) % 100) * 1.5,
})

for _ in range(WARMUP):
    df.rolling(WINDOW).var()
    df.rolling(WINDOW).std()
    df.rolling(WINDOW).sum()
    df.rolling(WINDOW).count()

times = []
for _ in range(ITERATIONS):
    t0 = time.perf_counter()
    df.rolling(WINDOW).var()
    df.rolling(WINDOW).std()
    df.rolling(WINDOW).sum()
    df.rolling(WINDOW).count()
    times.append((time.perf_counter() - t0) * 1000)

total_ms = sum(times)
mean_ms = total_ms / ITERATIONS
print(json.dumps({
    "function": "dataframe_rolling_var_std_sum_count",
    "mean_ms": round(mean_ms, 3),
    "iterations": ITERATIONS,
    "total_ms": round(total_ms, 3),
}))

File: benchmarks/tsb/bench_dataframe_rolling_var_std_sum_count.ts
/**
 * Benchmark: DataFrameRolling.var / std / sum / count — rolling aggregations on a 50k-row DataFrame.
 * Outputs JSON: {"function": "dataframe_rolling_var_std_sum_count", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { DataFrame } from "../../src/index.ts";

const SIZE = 50_000;
const WINDOW = 20;
const WARMUP = 5;
const ITERATIONS = 30;

const df = DataFrame.fromColumns({
  a: Array.from({ length: SIZE }, (_, i) => Math.sin(i * 0.01) * 100),
  b: Array.from({ length: SIZE }, (_, i) => Math.cos(i * 0.01) * 50),
  c: Array.from({ length: SIZE }, (_, i) => (i % 100) * 1.5),
});

for (let i = 0; i < WARMUP; i++) {
  df.rolling(WINDOW).var();
  df.rolling(WINDOW).std();
  df.rolling(WINDOW).sum();
  df.rolling(WINDOW).count();
}

const times: number[] = [];
for (let i = 0; i < ITERATIONS; i++) {
  const t0 = performance.now();
  df.rolling(WINDOW).var();
  df.rolling(WINDOW).std();
  df.rolling(WINDOW).sum();
  df.rolling(WINDOW).count();
  times.push(performance.now() - t0);
}

const totalMs = times.reduce((a, b) => a + b, 0);
const meanMs = totalMs / ITERATIONS;
console.log(
  JSON.stringify({
    function: "dataframe_rolling_var_std_sum_count",
    mean_ms: Math.round(meanMs * 1000) / 1000,
    iterations: ITERATIONS,
    total_ms: Math.round(totalMs * 1000) / 1000,
  }),
);
